### ISSUE TITLE
EXCEPTION NetworkConstructionDependencyLoopException: Error: There is a dependency loop on layer 'accum_att_weights'.

### ISSUE THREAD

**Issue by manish-kumar-garg** [rwth-i6/returnn-experiments#35]
Hi, I am getting NetworkConstructionDependencyLoopException while running 2019-asr-local-attention librispeech setup.
Full stdout [here](https://gist.github.com/manish-kumar-garg/ab616732fb286c61162b03db9732358a)

Few lines from where error starts:
```
EXCEPTION
NetworkConstructionDependencyLoopException: Error: There is a dependency loop on layer 'accum_att_weights'.
Construction stack (most recent first):
  accum_att_weights
  weight_feedback
  energy_in
  energy_tanh
  energy
  energy_reinterpreted
  att_weights
  att0
  att
  s
  readout_in
  readout
  output_prob

EXCEPTION
CannotHandleUndefinedSourcesException: 's_transformed': cannot handle undefined sources without defined out_type.
{'activation': None,
 'loss': None,
 'n_out': 1024,
 'network': <TFNetwork 'root/output:rec-subnet' parent_net=<TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'size_target': None,
 'sources': [None],
 'target': None,
 'with_bias': False}

Exception creating layer root/'output' of class RecLayer with opts:
```

**Comment by albertz** [comment]
Unfortunately the configs are slightly buggy, and do not work correctly with a more recent RETURNN version.
It should be easy to fix, though. Change this:

    "p_t_in": {"class": "eval", "from": "prev:att_weights", "eval": "tf.squeeze(tf.argmax(source(0), axis=1, output_type=tf.int32), axis=1)",
      "out_type": {"shape": (), "batch_dim_axis": 0, "dtype": "float32"}},

To:

    "p_t_in": {"class": "reduce", "from": "prev:att_weights", "mode": "argmax", "axis": "t"},

Also make sure that you use the latest RETURNN version.

**Comment by manish-kumar-garg** [comment]
After making this change with latest RETURNN version, 
Getting incorrect shapes [error](https://gist.github.com/manish-kumar-garg/7b017cfed5827a9083d5e16f7df43fda)

**Comment by manish-kumar-garg** [comment]
After making this change and using latest RETURNN version.
Getting incorrect shapes
```
TensorFlow exception: Incompatible shapes: [14,1,45] vs. [45,1,45]
	 [[node output/rec/att_weights/LogicalAnd_1 (defined at /home/ubuntu/rwth-i6/returnn-experiments/2018-asr-attention/librispeech/full-setup-attention/returnn/TFNetworkLayer.py:3159) ]]

```
 [stdout](https://gist.github.com/manish-kumar-garg/7b017cfed5827a9083d5e16f7df43fda)

**Comment by albertz** [comment]
Thank you. There was actually a real bug, which I fixed now (commit 7999f7430cb968).
The test `test_rec_layer_local_att_train_and_search` should cover this now.
Can you try again with latest Returnn?

**Comment by manish-kumar-garg** [comment]
It works now. Thanks!

**Comment by manish-kumar-garg** [comment]
@albertz , 
After making this change,
I can see that the model is not converging as expected.

Logs [here](https://gist.github.com/manish-kumar-garg/620f0fbf72029ff69de916e67ef23094)

Loss is always around 6k-7k

**Comment by albertz** [comment]
We did three kinds of experiments in the paper:
* Training with global soft attention, and then just importing it into this local soft attention config.
* Training with global soft attention, then importing it as local soft attention, and further training a bit.
* Training with local soft attention from scratch.

@Spotlight0xff are these the configs you used for importing the model, or from scratch training? If only for importing, can you also add the configs for the from-scratch training? (Or maybe just one reasonable/representative config for that.)

I remember that the from scratch training was quite unstable and needed some more tuning.

**Comment by manish-kumar-garg** [comment]
I have not tried that yet. Will try now.

**Comment by Spotlight0xff** [comment]
Hi,

The configs in `librispeech/` were all for pretrained (global attention) models (so case 1 and 2, but not 3.).
I just added a config for from-scratch training [here](https://github.com/rwth-i6/returnn-experiments/commit/da852ea7dfa50d0eaac5481a8e7a7ba1ccc5332b).
The issue with your model was probably the too small learning rate which was used when we retrained the initialized models.
